File: Medium/4020-lexicographically-smallest-permutation-greater-than-target/lexicographically-smallest-permutation-greater-than-target.py
class Solution:
    def solve(self, freq):
        str=""
        for i in range(26):
            for _ in range(freq[i]):
                str += chr(ord('a')+i)

        return str;
    
    def lexGreaterPermutation(self, s: str, t: str) -> str:
        n=len(s)

        freq=[0]*26
        for ch in s:
            freq[ord(ch)-ord('a')]+=1

        res=""
        prefix=""
        for ch in t:
            for c in range(ord(ch)+1, ord('z')+1):
                if freq[c-ord('a')]>0:
                    freq[c-ord('a')]-=1

                    suffix=self.solve(freq)
                    str=prefix+chr(c)+suffix

                    if res == "" or str<res:
                        res=str
                    
                    freq[c-ord('a')]+=1
                    break
                

            if freq[ord(ch)-ord('a')] == 0:
                return res

            freq[ord(ch)-ord('a')]-=1
            prefix += ch
        
        return res

File: 4020-lexicographically-smallest-permutation-greater-than-target/lexicographically-smallest-permutation-greater-than-target.java
class Solution {
    String solve(int[] freq){
        StringBuilder str = new StringBuilder();
        for (int i=0; i<26; i++){
            char c =(char)('a'+i);
            for(int j=0; j<freq[i]; j++){
                str.append(c);
            }
        }

        return str.toString();
    }

    public String lexGreaterPermutation(String s, String t){
        int n=s.length();
        int m=t.length();

        int[] freq=new int[26];
        for(char ch: s.toCharArray()){
            freq[ch-'a']++;
        }

        String res="";
        StringBuilder prefix=new StringBuilder();
        for(int i=0; i<m; i++){
            char ch = t.charAt(i);

            for(char c=(char)(ch+1); c<='z'; c++){
                if(freq[c-'a']>0){
                    freq[c-'a']--;

                    String suffix=solve(freq);
                    String str=prefix.toString()+c+suffix;
                    if (res.equals("") || str.compareTo(res)<0){
                        res=str;
                    }

                    freq[c-'a']++;
                    break;
                }
            }

            if(freq[ch-'a'] == 0)
                return res;

            freq[ch-'a']--;
            prefix.append(ch);
        }

        return res;
    }
}